packages/auth/src/grant/service.ts

```diff
@@ -142,12 +146,14 @@ async function rejectGrant(
   })
 }
 
-async function deleteGrant(
+async function revokeGrant(
   deps: ServiceDependencies,
   continueId: string
 ): Promise<boolean> {
-  const deletion = await Grant.query(deps.knex).delete().where({ continueId })
-  if (deletion === 0) {
+  const update = await Grant.query(deps.knex)
+    .patch({ state: GrantState.Revoked })
+    .where({ continueId })
+  if (update === 0) {
     deps.logger.info(
       `Could not find grant corresponding to continueId: ${continueId}`
     )
```

Comment:
Wasn’t sure about changing this one as well. After sitting with it a bit longer I think maybe we should continue hard-deleting here. Or, soft-deleting via a new `deletedAt` field (see note in PR body). This would mean there isnt parity between gql api and the rest api (one is delete, one is state update). Or change the http method from delete to post/put but probably not…

Currently in my PR the delete HTTP method will update the state and then we just handle grants with a `REVOKED` state differently depending on where we get them, which definitely seems wrong.

Reply:
A revoke grant REST endpoint should probably be a PATCH, not a DELETE. We actually call the DELETE endpoint "Cancel grant" [in the docs](https://docs.openpayments.guide/reference/delete-continue), which is for a client to withdraw his grant request or to surrender the grant. I think you are right in that in such a scenario, we shouldn't revoke but we should have a `deletedAt` field. And revocation should then only happen via the admin APIs because the user itself would revoke using his ASE's interface.

Reply:
After your comment and looking at the GNAP spec I think a `deletedAt` field for the DELETE method is what I'm going to do for this, but it seems like the spec has changed and our implementation doesn't align on a few things. Then again this is the first I've waded into the auth server and the GNAP spec so correct me if I'm wrong or missing some context.

Here is the specification for this DELETE route: https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol-14

Of note it specifies this route is for "Revoking a Grant Request" (previously "Canceling" like we use). This suggests there isn't the distinction between deleting and revoking like I thought. Not so simple to just revoke though because it specifies:
   - updating grant state to `finalized` (not `revoked`). Which doesn't exist in our code. It seems like the grant states are now (`processing`, `pending`, `approved`, `finalized`) instead of our (`pending`, `granted`, `revoked`, `rejected`). These cut across many operations and in terms of `revoked` and `finalized`, im not sure they are just interchangable.
   - should disable all token rotation and other token management functions. I think we are _effectively_ doing this because the tokens are on the grant that are deleted (so they cant be rotated). This is why I want to do a soft-delete instead of using grant state (`revoked` or `finalized`) but it doesn't seem like this is ultimately the right way to handle revocation.

Reply:
There may not be too much to make GNAPy after all. I noticed that we are deleting access and access token as well on grant deletion (via a postgres `ON DELETE CASCADE`). Which means I can't just draw the line at a soft delete of grants anyways. I ended up changing both the `DELETE /continue` and `revokeGrant` routes to update the grant state to `REVOKED` and handled or filtered out grants/access/access tokens with a revoked grant state on the read side. All the token and grant openapi requests should remain the same but we can now get the revoked grants via the gql admin api.

I still wonder about the states I see in the spec (`processing`, `pending`, `approved`, `finalized`) vs. what we have (`pending`, `granted`, `revoked`, `rejected`). Perhaps our `GrantStates` are not meant to reflect the states here: https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol-11`#section-1.5 For example, we have no concept of `finalized` but it was in version 11 of the spec. Still, the spec is pretty explicit about when to update the grant state to these values. We have `pending` and `approved` might map pretty clearly to `granted` but `finalized` doesn't really seem to have an analog.

Reply:
I think we should add the distinction between the `processing` and `pending` states. Respectively, one indicates the AS is determining what to do next with the grant and the other indicates that the grant is awaiting interaction.

As for `finalized`, I would consider `granted`, `revoked`, and `rejected` to be the AS's (granulized) concept of GNAP's `finalized`. It seems like the concept of a `finalized` state irrespective of _how_ a grant was finalized is all that matters for GNAP, but for our purposes we need more specific reasons why. I would think we should maybe modify the Grant model so that `granted/revoked/rejected` is replaced by `finalized` and add another field indicating how it was finalized. I think it would make this easier to follow as an implementation of GNAP that way, and we could keep information that we need.

I agree with @sabineschaller that this is a separate, larger issue that should be worked on separately from this, though.